storage: don't log benign queue errors

tbg opened the pull request:
This is a WIP that I'll have to complete when more pressing issues have been
addressed.

----

Queues can run into a variety of errors which are "harmless" in nature
(assuming they are not permanent).

Introduce a mechanism to detect such errors and avoid having them
increment the error metrics (and pollute the logs).

Touches #31373.

Release note: None

petermattis commented on pkg/storage/queue.go:
```diff
@@ -754,6 +754,35 @@ func (bq *baseQueue) processReplica(queueCtx context.Context, repl *Replica) err
+
+var _ benigner = benignError{}
+
+func (benignError) Benign() {}
```
I think `benignError` should implement `Cause()`.

petermattis reviewed (commented):
Approach looks good to me.

Be sure to wrap mark the error returned from `Replica.setPendingSnapshotIndex` as benign.

Also, my fingers want to spell `benign` as `begin`.

bdarnell commented on pkg/storage/queue.go:
```diff
@@ -754,6 +754,35 @@ func (bq *baseQueue) processReplica(queueCtx context.Context, repl *Replica) err
 	return nil
 }
 
+type benigner interface {
```
I'd call this interface `benignError` and the struct `benignWrapper`. (I can spell "benign" instead of "begin", but I can't verb it)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31728)
<!-- Reviewable:end -->

tbg reviewed (commented):
Ready for another look. I'm going to take this on a spin to see if it works in the real world.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-:-LPzmimF-ibaKhNvixSO:b-wlti2g)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/queue.go, line 757 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPXFYlFg1sbTQn4pr2Z-r1-757:-LPzIS-aBUycMIcqVMAF:b8cpf5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L757)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd call this interface `benignError` and the struct `benignWrapper`. (I can spell "benign" instead of "begin", but I can't verb it)
</blockquote></details>

Renamed to `benignI` (my brain wants `benignError` to be a struct).
___
*[pkg/storage/queue.go, line 767 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPXFYlFg1sbTQn4pr2Z-r1-767:-LPzTlvqCUp5RAyoaaJj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L767)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think `benignError` should implement `Cause()`.
</blockquote></details>

Done.
___
*[pkg/storage/queue.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPzTrAK7Q5mVOtp1Dm9:-LPzTrAL2di2T4eCTkTM:bwbb2vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L811)):*
> ```Go
> 		// Increment failures metric to capture all non-benign errors.
> 		if !benign {
> 			bq.failures.Inc(1)
> ```

I think I'd prefer to keep logging those as failures. Thoughts?


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-:-LQ-1En92Z_lSMujcojQ:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/queue.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPzTrAK7Q5mVOtp1Dm9:-LQ-0nkKBRZo3aahQNia:b-8ebarl) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L811)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think I'd prefer to keep logging those as failures. Thoughts?
</blockquote></details>

They aren't successes. They aren't really failures either. Perhaps there is a 3rd category to place them under. I guess I don't feel strongly one way or the other.

Update: looking at this a bit more, I think sticking with the failure metric is the right way to go. In particular, if the remote declines a snapshot it does seem like a replicate queue failure, though the error is benign and doesn't need to be logged.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/queue.go, line 757 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPXFYlFg1sbTQn4pr2Z-r1-757:-LQ-ynU-EsQk5qocv3az:bm74cp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L757)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Renamed to `benignI` (my brain wants `benignError` to be a struct).
</blockquote></details>

Actually, do we even need the interface? When would we implement the `Benign` method on another error type instead of just wrapping another error in a `benignError`? 
___
*[pkg/storage/queue.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPzTrAK7Q5mVOtp1Dm9:-LQ-ybp670ARkTUpOUcp:b-hlj7uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L811)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

They aren't successes. They aren't really failures either. Perhaps there is a 3rd category to place them under. I guess I don't feel strongly one way or the other.

Update: looking at this a bit more, I think sticking with the failure metric is the right way to go. In particular, if the remote declines a snapshot it does seem like a replicate queue failure, though the error is benign and doesn't need to be logged.
</blockquote></details>

I'm OK with marking them as failures but I think they're more like the cases where ShouldQueue returned false and we decided not to process the replica (which I don't think increments any stats). 


<!-- Sent from Reviewable.io -->

tbg committed:
causer: provide method for walking an error chain

Release note: None

tbg committed:
storage: don't log benign queue errors

Queues can run into a variety of errors which are "harmless" in nature
(assuming they are not permanent).

Introduce a mechanism to detect such errors and avoid having them
increment the error metrics (and pollute the logs).

Touches #31373.

Release note: None

tbg committed:
storage: harden purgatory error check

I didn't notice a concrete problem here, but it's more robust to walk
the error chain.

Release note: None

tbg committed:
storage: harden admin split retryable errors

If either of the two special-cased errors ever implemented Causer, the
old code would jump right past them.

Release note: None

tbg reviewed (commented):
PTAL, a bit of cleanup has happened. I'm going to kick off another run of a large enough import test to weed out any problems (changing anything about how errors are wrapped often tickles untested code paths that are required to keep scatter from failing eagerly etc).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-:-LQ9z9cp7kLcs2q4fX5I:b5per8h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/queue.go, line 757 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPXFYlFg1sbTQn4pr2Z-r1-757:-LQ9nTrr1UOI2OaTMPoD:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L757)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Actually, do we even need the interface? When would we implement the `Benign` method on another error type instead of just wrapping another error in a `benignError`? 
</blockquote></details>

Good point. Done.
___
*[pkg/storage/queue.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LPzTrAK7Q5mVOtp1Dm9:-LQ9pGJcAbUJIoZz6m8A:bqeq3rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/831c97323336cdf7b46cad21c79b118807b2a1b3/pkg/storage/queue.go#L811)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm OK with marking them as failures but I think they're more like the cases where ShouldQueue returned false and we decided not to process the replica (which I don't think increments any stats). 
</blockquote></details>

I'm going to keep tracking them as failures for now. I expect that over time we'll want to track them separately (`bq.benign.Inc(1)`) because I expect our testing down the road to want to assert that there aren't ever any queue failures in normal operation. We're still a far cry from that right now though.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-:-LQAFofvDwYb2tQ7x3ZO:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

bors r=bdarnell

tschottdorf commented:
bors r=bdarnell

retrying since the CI failure was likely fixed by https://github.com/cockroachdb/roachprod/pull/234.

tbg commented:
@bdarnell wasn't planning on backporting this, but let me know if you'd like it on release-2.1.

petermattis reviewed (commented):
I'm fine with a backport. 2.2 is a long way off and it will be nice to omit these benign errors from the logs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-:-LRS9eO45Xhrmx3h-lBD:b-qtja8n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/replica_command.go, line 176 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31728#-LRS9IUh8SLFLMRn6E8c:-LRS9IUh8SLFLMRn6E8d:b-2zfpgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/acb950a3c0761411c1c3a4cec808089b8be6c946/pkg/storage/replica_command.go#L176)):*
> ```Go
> 		// On seeing a ConditionFailedError or an AmbiguousResultError, retry
> 		// the command with the updated descriptor.
> 		if retry := causer.Visit(lastErr, func(err error) bool {
> ```

Was this part of the change motivated by something? Is it a bug fix? `roachpb.{ConditionalFailed,AmbiguousResult}Error` do not implement the `Causer` interface, so on the surface this seems to be the same as the old code?


<!-- Sent from Reviewable.io -->

bdarnell commented:
I'd vote against backporting but don't feel strongly.

tbg commented:
@petermattis 

> Was this part of the change motivated by something? Is it a bug fix?

No, just future proofing.

tbg commented:
I won't backport this.
